pokemon.go

```diff
@@ -53,6 +67,137 @@ func GetPokemon(natdex uint16) Pokemon {
 	panic(fmt.Sprintf("unknown Pokedex number %v\n", natdex))
 }
 
+// Creates a new Pokemon given its national dex number
+func GeneratePokemon(args ...interface{}) Pokemon {
+	// Mandatory parameters
+	var natdex uint16
+
+	// Optional parameters
+	var level uint8 = 1
+	var totalExp uint // used to compute level if provided
+	var ivs [6]uint8 = [6]uint8{0, 0, 0, 0, 0, 0}
+	var evs [6]uint8 = [6]uint8{0, 0, 0, 0, 0, 0}
+	var nature *Nature = GetNatureTable()["hardy"] // this nature provides no bonuses/debuffs
+
+	if len(args) < 2 {
+		panic("Not enough parameters.")
+	}
+
+	for i, p := range args {
+		switch i {
+		case 0: // natdex
+			param, ok := p.(uint16)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type uint16, got %v", i, reflect.TypeOf(p)))
+			}
+			natdex = param
+		case 1: // level or total exp gained
+			switch p.(type) {
+			case uint8: // level
+				level = p.(uint8)
+				totalExp = computeExpFromLevel(level, getGrowthRateFromDexNumber(natdex))
+			case uint: // total experience
+				totalExp = p.(uint)
+				level = computeLevelFromExp(totalExp, getGrowthRateFromDexNumber(natdex))
+			default:
+				panic(fmt.Sprintf("Parameter %d: expected type uint8 or uint, got %v", i, reflect.TypeOf(p)))
+			}
+		case 2: // ivs
+			param, ok := p.([6]uint8)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type [6]uint8, got %v", i, reflect.TypeOf(p)))
+			}
+			ivs = param
+		case 3: //evs
+			param, ok := p.([6]uint8)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type [6]uint8, got %v", i, reflect.TypeOf(p)))
+			}
+			evs = param
+		case 4: // nature
+			param, ok := p.(*Nature)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type Nature, got %v", i, reflect.TypeOf(p)))
+			}
+			nature = param
+		}
+	}
+
+	for _, p := range ALL_POKEMON {
+		if p.NatDex == natdex {
+			p.Level = level
+			p.TotalExperience = totalExp
+			p.IVs = ivs
+			p.EVs = evs
+			p.Nature = nature
+			p.computeStats()
+			return p
+		}
+	}
+
+	// Not exactly the best way to handle this
+	panic(fmt.Sprintf("unknown Pokedex number %v\n", natdex))
+}
+
+func computeLevelFromExp(exp uint, growth_rate int) uint8 {
+	experience_csv := getCsvReader("data/experience.csv")
+
+	var lastExp uint = 0
+	for {
+		record, err := experience_csv.Read()
+
+		if err == io.EOF {
+			break
+		}
+
+		nextExp := uint(parseInt(record[2]))
+		if parseInt(record[0]) == growth_rate {
+			if exp > lastExp && exp <= nextExp {
+				return uint8(parseInt(record[1]))
+			}
+		}
+
+		lastExp = nextExp
+	}
+	panic("There was a problem computing the level.")
+}
+
+func computeExpFromLevel(level uint8, growth_rate int) uint {
+	experience_csv := getCsvReader("data/experience.csv")
```

Comment:
Reading from `data/` at runtime is not allowed. All info must be included in the compiled binary.